### DISCUSSION TITLE
Calibration of Current Transformer: do you compensate for small offset values at no load?

### DISCUSSION THREAD

**Discussion by mbeaudoin** [CircuitSetup/Expandable-6-Channel-ESP32-Energy-Meter#179] [Q&A] [answered]
I am in the process of calibrating my current transformers. (a pair of 200A/50mA SCT-024 and some 100A/50ma SCT-013-000)
Using a resistive load drawing about 7 Amps, I can easily adjust the individual transformer 'gain_ct' factors to match the reading from my Fluke current clamp. So far, so good.

But when shutting down the resistive load while still keeping the current transformers in-place, so with zero Amp flowing through the circuit, the 6 Channel main board is still reporting some small current and power values, like 0.14A for the current, and a few hundredth of Watt for the power. I would expect to read 0 Amps in that case. I guess I am rather hitting the measurement precision of the ATM90E32 chips?

I tried to compensate for those tiny offsets using an esphome filter offset, but it is like chasing a moving target, the current and power offsets remains tiny, but I cannot compensate for them completely. 

Any comments would be  appreciated.

**Comment by SzymonSlupik** [comment]
Same on my end. 
It sounds kind of similar to this one: [https://forum.microchip.com/s/topic/a5C3l000000Me89EAC/t384316](https://forum.microchip.com/s/topic/a5C3l000000Me89EAC/t384316).
But I have no idea if this is a correct observation and/or how to fix the non-zero offset values in ESP Home...

**Reply by SzymonSlupik** [reply]
after some observations I see these offset values are different after each reset of the ESP32. So it may be a bug in the calibration procedure (if there is such thing). I tried correcting the offsets by using the 
`filters: 
          - offset: `
but it did not help - after ESP32 reset I have different offsets (and sometimes negative)

**Reply by mbeaudoin** [reply]
Same observations here. The offsets are changing on resets, so any offset filter is of no use. 
I ended up clamping the tiny negative power values to zero, and simply moved on. 

I also saw the thread you've mentioned. The CircuitSetup boards do have a 10.1uF decoupling capacitor on the VRef pin. However, the ATM90E32AS chip docs is asking instead for a 4.7uF capacitor. Not sure why they chose to go with a bigger capacitor value.
https://ww1.microchip.com/downloads/en/devicedoc/Atmel-46003-SE-M90E32AS-Datasheet.pdf

This really looks like an uninitialized variable in the ATM90E32AS  esphome driver source code, but I could not find such basic error when reading the source code.

**Reply by SzymonSlupik** [reply]
I've added an add-on board and now it seems to me it is more hardware-related problem or combination of hardware and insufficient calibration.

The add-on board shows very small offsets (0.001A) or no offsets at all.

One other hardware difference I have is the add-on board has screw terminals while the main board has 3.5mm jacks. So it may also be that the jack connectors are introducing something? Or some components misplaced on the board?

![image](https://github.com/user-attachments/assets/5064e1cd-da1b-461b-a9d2-189f61a36c34)

**Reply by CircuitSetup** [reply]
> Same observations here. The offsets are changing on resets, so any offset filter is of no use. I ended up clamping the tiny negative power values to zero, and simply moved on.

The way that offsets are calculated on startup isn't ideal. See here: https://github.com/esphome/esphome/blob/1f3754684adccccc54a4795d8a685d13ba59e352/esphome/components/atm90e32/atm90e32.cpp#L170

According to the application note (https://ww1.microchip.com/downloads/aemDocuments/documents/OTH/ApplicationNotes/ApplicationNotes/Atmel-46103-SE-M90E32AS-ApplicationNote.pdf page 21) offsets should be calculated with transformers reading 0

So, depending on what is flowing through the CT during startup could change the offset calculation. Unplugging the CTs during startup would help, but those offset values would be lost if the meter lost power. That being said, the offset values are rarely significant enough to cause much difference. In fact, when I tested this, I don't remember the offset registers making any difference at all. 

Perhaps this can be changed in the ESPHome code so that in can be saved as a variable, or just turned off completely. That would eliminate the question of offsets.

> I also saw the thread you've mentioned. The CircuitSetup boards do have a 10.1uF decoupling capacitor on the VRef pin. However, the ATM90E32AS chip docs is asking instead for a 4.7uF capacitor.

I don't remember exactly why this was done, but I don't believe it will matter in terms of incorrect current readings, as the difference between 10uF and 4.7uF would be filtering out lower transient frequencies more effectively. In the thread referenced it looks like there was no bypass cap at all, which could cause significant noise on the vref. 

Overall, the accuracy of the CT's (+-1%), and how accurately they're calibrated to the meter, will cause the most variation. I realize this can be frustrating, though, especially when the readings don't appear to be linear.  

> One other hardware difference I have is the add-on board has screw terminals while the main board has 3.5mm jacks. So it may also be that the jack connectors are introducing something? Or some components misplaced on the board?

The connectors are all connected exactly the same way. The only other difference I can think of would be when the boards where assembled. There may be slightly different burden resistors used between the boards. Even so, they should be within 0.1% if they're both newer meters (v1.4 rev1). 

Are all CTs exactly the same?

**Reply by SzymonSlupik** [reply]
Also reading this: [https://github.com/CircuitSetup/Expandable-6-Channel-ESP32-Energy-Meter](https://github.com/CircuitSetup/Expandable-6-Channel-ESP32-Energy-Meter):

> Q: I'm getting a small negative value when there is no load at all, but a positive value otherwise. What is going on?
>
>A: This is caused by variances in resistors and current transformers. You can either calibrate the current transformers to the meter, or add this lambda section to only allow positive values for a current channel:
```

  - platform: template
    name: ${disp_name} CT1 Watts Positive
    id: ct1WattsPositive
    lambda: |-
      if (id(ct1Watts).state < 0) {
        return 0;
      } else {
        return id(ct1Watts).state ;
      }
    accuracy_decimals: 2
    unit_of_measurement: W
    icon: "mdi:flash-circle"
    update_interval: ${update_time}
```

>Then for your total watts calculation, use ct1WattsPositive

It may be that indeed [_this is caused by variances in resistors_].

But OTOH the fact the offset changes after every reboot suggests the calibration somehow fails on that particular hardware (the particular batch of main boards?).

**Reply by CircuitSetup** [reply]
> But OTOH the fact the offset changes after every reboot suggests the calibration somehow fails on that particular hardware (the particular batch of main boards?).

Are you observing the value of the offset registers changing here?

**Comment by SzymonSlupik** [comment]
Hi CircutSetup - thanks for your input.

> _So, depending on what is flowing through the CT during startup could change the offset calculation. Unplugging the CTs during startup would help, but those offset values would be lost if the meter lost power. That being said, the offset values are rarely significant enough to cause much difference. In fact, when I tested this, I don't remember the offset registers making any difference at all._

Well in my case there is NOTHING plugged in the CT sockets. So definitely no current is flowing (other that maybe induced on the PCB?).

I clearly see a big difference in those offsets between the main board and the expansion boards.

**Reply by CircuitSetup** [reply]
Do you have a multimeter handy? I'm curious what the burden resistors (the ones right next to the CT plugs) are reading on your main board vs add-on board. They should be 22 ohms.

**Reply by SzymonSlupik** [reply]
John, yes, all are 22 ohms on both boards (i measured that between the plug terminals).

**Reply by CircuitSetup** [reply]
Does you meter read out tenths? It just may be the difference between the resistors used between manufacturing runs. 

Also, have the CTs been calibrated or the calibration values changed at all?

**Reply by SzymonSlupik** [reply]
All the resistors show between 21.4 and 21.6 ohm. So within 1% from each other.

The issue is these offsets change (are different) after each reset. So I'm not sure how CT calibration has anything to do with this, especially as the offsets fluctuate (after each reset) with no CTs plugged in.

To me it has something to do with the calibration procedure which does not do a good job on reset. For some reason it seems to work better on the add-on board channels. BTW I'm in Europe (50Hz) so maybe it matters?

I saw in the source code it takes 5 measurements and then averages them. But I don't know how that works (it takes the lowest 16 bits and writes them to the chip, but not sure what is the meaning of that?).

**Reply by macduff89** [reply]
Hi, just wanted to note I'm observing a similar thing - some sensors show an offset current up to 0.04 A when it should be 0. Whether there is nothing plugged into the jack, or there is a CT plugged in but no current flowing makes no difference. And the sensors with offset change after every reboot, although in my case it doesn't seem to matter if it's on the main board or add-on board. However, strangely these current errors do not correspond with power readings. For example in the history graph below Air Con Current shows a constant 0.04A, which should be about 10W (I'm in Australia using 240V) but it reads 0.01W.

![Screenshot 2024-08-02 100514](https://github.com/user-attachments/assets/fdc01718-8af3-4c49-b170-b3969d9a6285)

**Comment by SzymonSlupik** [comment]
I think we have identified the root cause for the calibration offsets. See https://github.com/CircuitSetup/Expandable-6-Channel-ESP32-Energy-Meter/issues/88#issuecomment-2254141699. 
I'm not sure though why your wattage is not a multiply the current by the voltage though...

**Comment by descipher** [comment, marked as answer]
https://github.com/esphome/esphome/pull/7228

**Comment by SzymonSlupik** [comment]
Jut wanted to say THANK YOU for resolving this issue. I finally found some hobby time to put everything together. Things work great, I did not need to run the manual calibration on my meter as the offsets were really negligible.
Happy New Year!
[https://headworx.slupik.com/2024/12/home-energy-monitoring.html](url)
